### DISCUSSION TITLE
Actions of invocation cannot be extracted

### DISCUSSION THREAD

**Discussion by baeharam** [statelyai/xstate#1706] [Q&A] [answered]
I can extract `actions` from `states` into `actions` section of machine. However, `actions` of `invocation` cannot be extracted into `actions` section because of type error in Typescript. Is this intention? if not, how can I solve this?

In [Fully reproducible codesandbox](https://codesandbox.io/s/extract-invoking-actions-qpeff?file=/src/counterMachine.ts), you can check by investigating `countMachine.ts`. There is type error in `actions` section related to action of invocation.

Thx.

**Comment by davidkpiano** [comment, marked as answer]
This is a known limitation (which might be fixed soon). You have to assume that actions in `options.actions: { ... }` can be called from any state, with any event. That's why the type has to be the full `TEvent`, and not a restricted event type.

**Reply by baeharam** [reply]
Then, do I have to use `createMachine<SomeContext, any, any>` to fix it now?

**Reply by davidkpiano** [reply]
No, just add `DoneInvokeEvent<...>` to your event types:

```ts
export type CounterEvent =
  | { type: "increase" }
  | { type: "decrease" }
  | { type: "increaseByInvoking" }
  | DoneInvokeEvent<number>;
```

And check for that event:

```ts
  {
    actions: {
      // ...
      onDoneAction: assign((context, event) => {
        if (!('data' in event)) { return; }

        context.count = event.data;
      })
    }
  }
```

Then it's totally type-safe. [CodeSandbox example](https://codesandbox.io/s/extract-invoking-actions-forked-hujft?file=/src/counterMachine.ts)
